apps/admin/backend/src/server.ts

```diff
@@ -192,11 +206,131 @@ function buildApi({
+        )
+        .map((cvr) => ({
+          ...cvr,
+          // Strip out ballot images to keep the response size low, since
+          // they're not needed client-side.
+          _ballotImages: undefined,
+        }));
+    },
+
+    async addCastVoteRecordFile({
+      path,
+    }: {
+      path: string;
+    }): Promise<AddCastVoteRecordFileResult> {
+      let fileStat: Stats;
+      try {
+        fileStat = await fs.stat(path);
+      } catch {
+        return err({
+          type: 'invalid-file',
+          userFriendlyMessage:
+            'The selected file could not be opened by the system.',
+        });
+      }
+
+      // try to get the exported timestamp from the filename
+      let exportedTimestamp: Date = fileStat.mtime;
+      try {
+        const parsedFileInfo = parseCvrFileInfoFromFilename(basename(path));
+        if (parsedFileInfo) {
+          exportedTimestamp = parsedFileInfo.timestamp;
+        }
+      } catch {
+        // file name was not in standard format, we'll try to import anyway
+      }
+
+      const addFileResult = await store.addCastVoteRecordFile({
+        electionId: loadCurrentElectionIdOrThrow(workspace),
+        filePath: path,
+        originalFilename: basename(path),
+        exportedTimestamp: exportedTimestamp.toISOString(),
+      });
+
+      const userRole = assertDefined(await getUserRole());
+      const filename = basename(path);
+      if (addFileResult.isErr()) {
+        const errorMessage = addFileResult.err().userFriendlyMessage;
+        await logger.log(LogEventId.CvrLoaded, userRole, {
+          message: `Failed to load CVR file: ${errorMessage}`,
+          disposition: 'failure',
+          filename,
+          error: errorMessage,
+          result: 'File not loaded, error shown to user.',
+        });
+        return err({ type: 'invalid-record', ...addFileResult.err() });
+      }
+
+      if (addFileResult.ok().wasExistingFile) {
+        // log failure if the file was a duplicate
+        await logger.log(LogEventId.CvrLoaded, userRole, {
+          message:
+            'CVR file was not loaded as it is a duplicate of a previously loaded file.',
+          disposition: 'failure',
+          filename,
+          result: 'File not loaded, error shown to user.',
+        });
+      } else {
+        // log success otherwise
+        await logger.log(LogEventId.CvrLoaded, userRole, {
+          message: 'CVR file successfully loaded.',
+          disposition: 'success',
+          filename,
+          numberOfBallotsImported: addFileResult.ok().newlyAdded,
+          duplicateBallotsIgnored: addFileResult.ok().alreadyPresent,
+        });
+      }
+
+      return addFileResult;
+    },
+
+    clearCastVoteRecordFiles() {
+      const electionId = loadCurrentElectionIdOrThrow(workspace);
+      store.deleteCastVoteRecordFiles(electionId);
+      store.setElectionResultsOfficial(electionId, false);
```

Comment:
Unrelated to the choices or changes in this PR, but currently when we clear the cast vote record files, we chance an election from "marked official" to "not marked official." Haven't thought through this too much - is it a potential security loophole? cc @mattroe @benadida? We could instead do it only in "Test Ballot Mode."

Reply:
I don't see a problem.